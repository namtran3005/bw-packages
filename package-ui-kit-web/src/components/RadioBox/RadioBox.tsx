import * as React from 'react';
import styled from 'styled-components';
import { Radio as MuiRadio } from '@material-ui/core';
import { RadioProps as MuiRadioProps } from '@material-ui/core/Radio';
import {
  Colors,
  ColorsType,
  Spacing,
  Transitions,
} from '@bitwala-cryptobank-squad/package-theme';
import { Typography } from '../..';

interface RootComponentProps {
  foreground?: ColorsType;
  checked?: boolean | string;
  variant?: 'hollow' | 'solid';
}

interface Props {
  innerRef?: any; // eslint-disable-line @typescript-eslint/no-explicit-any
}

type RadioBoxProps = RootComponentProps & MuiRadioProps & Props;

const RootRadioComponent: React.FunctionComponent<RadioBoxProps> = (props) => (
  <MuiRadio {...props} />
);

const defaultForeground = 'primaryLilac';
const borderWidth = 1.4;

const RootRadio = styled(RootRadioComponent)`
  &&& {
    /* hide radio button while preserving focus availability */
    height: 0;
    overflow: hidden;
    position: absolute;
    opacity: 0;
  }
`;

const checkedBorderWidth = Spacing.unit / 4;

const Root: React.FC = styled.label.withConfig({
  shouldForwardProp: (prop) =>
    !['foreground', 'variant', 'checked'].includes(prop),
})<RootComponentProps>`
  display: inline-flex;
  cursor: pointer;
  background: ${({ variant }) =>
    variant === 'hollow' ? Colors.get('white') : Colors.get('primaryLilac')};
  border: ${({ variant }) =>
    variant === 'hollow'
      ? `${borderWidth}px solid ${Colors.get('text')}`
      : Colors.get('primaryLilac')};
  padding: ${Spacing.unit * 4}px ${Spacing.unit * 5}px;
  margin-right: ${Spacing.unit}px;
  transition: ${Transitions.duration.quick}ms ${Transitions.easing.swiftOut};
  box-shadow: 0 0 0 0 ${Colors.getWithAlpha(defaultForeground, 0.2)},
    inset 0 0 0 ${(props) => (props.checked ? checkedBorderWidth : 0)}px
      ${Colors.get(defaultForeground)};
  height: 48px;
  padding: 0px 18px;
  border-radius: 32px;
  align-items: center;
  vertical-align: middle;
  &:last-of-type {
    margin-right: 0;
  }

  &:focus-within {
    box-shadow: 0 0 0 ${checkedBorderWidth * 2}px
        ${Colors.getWithAlpha(defaultForeground, 0.2)},
      inset 0 0 0 ${(props) => (props.checked ? checkedBorderWidth : 0)}px
        ${Colors.get(defaultForeground)};
  }
`;

const RadioBox: React.FunctionComponent<RadioBoxProps> = ({
  children,
  ...cleanProps
}) => (
  <Root {...cleanProps}>
    <RootRadio {...cleanProps} />
    <Typography.ButtonL>{children}</Typography.ButtonL>
  </Root>
);

export { RadioBox, RadioBoxProps };
