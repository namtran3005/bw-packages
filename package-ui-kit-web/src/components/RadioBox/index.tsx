export { RadioBox, RadioBoxProps } from './RadioBox';
