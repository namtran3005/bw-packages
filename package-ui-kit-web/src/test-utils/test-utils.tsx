import {
  render as rtlRender,
  waitFor,
  RenderOptions as RTLRenderOptions,
} from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import userEvent from '@testing-library/user-event';
import { WebAppThemeProvider } from '../theme';
import '@testing-library/jest-dom';
import '@testing-library/jest-dom/extend-expect';

type RenderOptions = RTLRenderOptions;

function renderWithTestHarness(
  ui: React.ReactElement,
  renderOptions: RenderOptions = {}
) {
  function Wrapper({ children }: any) {
    return (
      <MemoryRouter>
        <WebAppThemeProvider>{children}</WebAppThemeProvider>
      </MemoryRouter>
    );
  }

  return rtlRender(ui, {
    wrapper: Wrapper,
    ...renderOptions,
  });
}

// Use if you're in a hurry (better than nothing)
// be sure to verify no memory leak from implicit
// waitFor by adding these flags to jest: --runInBand --detectOpenHandles
async function renderAct(ui: React.ReactElement, options: RenderOptions = {}) {
  const view = renderWithTestHarness(ui, options);
  return await waitFor(() => view, {
    timeout: 5000,
    onTimeout: (e) => {
      // eslint-disable-next-line no-console
      console.error(e);
      return e;
    },
  });
}

export * from '@testing-library/react';

export { renderWithTestHarness, renderAct, userEvent };
