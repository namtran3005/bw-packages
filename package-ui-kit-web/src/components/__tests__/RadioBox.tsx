import * as React from 'react';

import { renderWithTestHarness } from '../../test-utils/test-utils';
import { RadioBox } from '../RadioBox';

const defaultProps = {};

describe('RadioBox', () => {
  it('should render radio box', () => {
    const { baseElement } = renderWithTestHarness(
      <RadioBox {...defaultProps} />
    );
    expect(baseElement).toMatchSnapshot();
  });
  it('should render checked radio box', () => {
    const { baseElement } = renderWithTestHarness(
      <RadioBox checked {...defaultProps} />
    );
    expect(baseElement).toMatchSnapshot();
  });
});
