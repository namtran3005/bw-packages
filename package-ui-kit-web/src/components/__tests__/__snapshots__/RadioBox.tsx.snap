// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RadioBox should render checked radio box 1`] = `
<body>
  .c2 {
  text-align: left;
}

.c2.c2.c2 {
  font-weight: 500;
  font-size: 1rem;
  color: #2C232E;
  font-family: Inter;
  font-style: normal;
  white-space: inherit;
  -webkit-text-decoration: inherit;
  text-decoration: inherit;
  word-wrap: break-word;
}

.c2.c2.c2:hover {
  background: initial;
}

.c2.c2.c2:active {
  background: initial;
}

.c1.c1.c1 {
  height: 0;
  overflow: hidden;
  position: absolute;
  opacity: 0;
}

.c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  cursor: pointer;
  background: #BEAAFF;
  border: #BEAAFF;
  padding: 32px 40px;
  margin-right: 8px;
  -webkit-transition: 150ms cubic-bezier(.62,.28,.23,.99);
  transition: 150ms cubic-bezier(.62,.28,.23,.99);
  box-shadow: 0 0 0 0 rgba(190,170,255,0.2), inset 0 0 0 2px #BEAAFF;
  height: 48px;
  padding: 0px 18px;
  border-radius: 32px;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  vertical-align: middle;
}

.c0:last-of-type {
  margin-right: 0;
}

.c0:focus-within {
  box-shadow: 0 0 0 4px rgba(190,170,255,0.2), inset 0 0 0 2px #BEAAFF;
}

<div>
    <label
      class="c0"
    >
      <span
        aria-disabled="false"
        class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-8 MuiRadio-root MuiRadio-colorSecondary c1 PrivateSwitchBase-checked-9 Mui-checked MuiIconButton-colorSecondary"
      >
        <span
          class="MuiIconButton-label"
        >
          <input
            checked=""
            class="PrivateSwitchBase-input-11"
            type="radio"
            value=""
          />
          <div
            class="PrivateRadioButtonIcon-root-12 PrivateRadioButtonIcon-checked-14"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
              />
            </svg>
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root PrivateRadioButtonIcon-layer-13"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M8.465 8.465C9.37 7.56 10.62 7 12 7C14.76 7 17 9.24 17 12C17 13.38 16.44 14.63 15.535 15.535C14.63 16.44 13.38 17 12 17C9.24 17 7 14.76 7 12C7 10.62 7.56 9.37 8.465 8.465Z"
              />
            </svg>
          </div>
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </span>
      <span
        class="MuiTypography-root c2 MuiTypography-button"
      />
    </label>
  </div>
</body>
`;

exports[`RadioBox should render radio box 1`] = `
.c2 {
  text-align: left;
}

.c2.c2.c2 {
  font-weight: 500;
  font-size: 1rem;
  color: #2C232E;
  font-family: Inter;
  font-style: normal;
  white-space: inherit;
  -webkit-text-decoration: inherit;
  text-decoration: inherit;
  word-wrap: break-word;
}

.c2.c2.c2:hover {
  background: initial;
}

.c2.c2.c2:active {
  background: initial;
}

.c1.c1.c1 {
  height: 0;
  overflow: hidden;
  position: absolute;
  opacity: 0;
}

.c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  cursor: pointer;
  background: #BEAAFF;
  border: #BEAAFF;
  padding: 32px 40px;
  margin-right: 8px;
  -webkit-transition: 150ms cubic-bezier(.62,.28,.23,.99);
  transition: 150ms cubic-bezier(.62,.28,.23,.99);
  box-shadow: 0 0 0 0 rgba(190,170,255,0.2), inset 0 0 0 0px #BEAAFF;
  height: 48px;
  padding: 0px 18px;
  border-radius: 32px;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  vertical-align: middle;
}

.c0:last-of-type {
  margin-right: 0;
}

.c0:focus-within {
  box-shadow: 0 0 0 4px rgba(190,170,255,0.2), inset 0 0 0 0px #BEAAFF;
}

<body>
  <div>
    <label
      class="c0"
    >
      <span
        aria-disabled="false"
        class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-1 MuiRadio-root MuiRadio-colorSecondary c1 MuiIconButton-colorSecondary"
      >
        <span
          class="MuiIconButton-label"
        >
          <input
            class="PrivateSwitchBase-input-4"
            type="radio"
            value=""
          />
          <div
            class="PrivateRadioButtonIcon-root-5"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
              />
            </svg>
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root PrivateRadioButtonIcon-layer-6"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M8.465 8.465C9.37 7.56 10.62 7 12 7C14.76 7 17 9.24 17 12C17 13.38 16.44 14.63 15.535 15.535C14.63 16.44 13.38 17 12 17C9.24 17 7 14.76 7 12C7 10.62 7.56 9.37 8.465 8.465Z"
              />
            </svg>
          </div>
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </span>
      <span
        class="MuiTypography-root c2 MuiTypography-button"
      />
    </label>
  </div>
</body>
`;
